_user/rotrectangle/main.c

```diff
@@ -0,0 +1,277 @@
+	x0 = (unsigned int)(xc - ((sin(M_PI / 2 - alfa - atan(b / a))) * sqrt(pow(a, 2.0) + pow(b, 2.0))) / 2.0);
+	y0 = (unsigned int)(yc - ((cos(M_PI / 2 - alfa - atan(b / a))) * sqrt(pow(a, 2.0) + pow(b, 2.0))) / 2.0);
+
+	x = x0;
+	y = y0;
+	dx = (self->a) * cos(alfa);
+	dy = (self->a) * sin(alfa);
+	graph_line(graphp, x, y, (int)dx, (int)dy, 1, self->color, GRAPH_QUEUE_HIGH);
+
+	x = x0 + dx;
+	y = y0 + dy;
+	dx = (self->b) * cos(alfa + (0.5 * M_PI));
+	dy = (self->b) * sin(alfa + (0.5 * M_PI));
+	graph_line(graphp, x, y, (int)dx, (int)dy, 1, self->color, GRAPH_QUEUE_HIGH);
+
+	x = x0;
+	y = y0;
+	dx = (self->b) * cos(alfa + (0.5 * M_PI));
+	dy = (self->b) * sin(alfa + (0.5 * M_PI));
+	graph_line(graphp, x, y, (int)dx, (int)dy, 1, self->color, GRAPH_QUEUE_HIGH);
+
+	x = x0 + dx;
+	y = y0 + dy;
+	dx = (self->a) * cos(alfa);
+	dy = (self->a) * sin(alfa);
+	graph_line(graphp, x, y, (int)dx, (int)dy, 1, self->color, GRAPH_QUEUE_HIGH);
+
+	return 0;
+}
+
+
+static int rotrectangle_save(rotrectangle_t *self)
+{
+	self->prev.xc = self->rec.xc;
+	self->prev.yc = self->rec.yc;
+	self->prev.a = self->rec.a;
+	self->prev.b = self->rec.b;
+	self->prev.color = 0;
+
+	return 0;
+}
+
+
+static int rotrectangle_rotating(rotrectangle_t *self, graph_t *graphp)
+{
+	unsigned int alfa, ret;
+	ret = 0;
+	while (self->stop == 0) {
+		for (alfa = 0; alfa < 360; alfa += self->speed) {
+			if (self->stop != 0)
+				break;
+			ret = rotrectangle_print(&(self->rec), graphp, alfa);
+			/* filling rectangle functions are commented out, because of problem with screen refreshing, it's demo without it */
+			/* graph_fill(graphp, self->rec.xc, self->rec.yc, self->rec.color, GRAPH_FILL_FLOOD, GRAPH_QUEUE_HIGH); */
+			ret = rotrectangle_save(self);
+			usleep(10000); /* min value */
+			ret = rotrectangle_print(&(self->prev), graphp, alfa);
+			/* graph_fill(graphp, self->rec.xc, self->rec.yc, 0, GRAPH_FILL_FLOOD, GRAPH_QUEUE_HIGH); */
+		}
+	}
+
+	return ret;
+}
+
+
+static int rotrectangle_startpanel(rotrectangle_t *self)
+{
+	printf("*** Rotating rectangle program ***\n");
+	printf("  User interface:\n");
+	printf("    )up arrow - move up\n");
+	printf("    )down arrow - move down\n");
+	printf("    )right arrow - move right\n");
+	printf("    )left arrow - move left\n");
+	printf("    )'+' - increase rotation speed\n");
+	printf("    )'-' - decrease rotation speed\n");
+	printf("    )'q' - exit\n");
+	printf("  Press enter to start: \n");
+	getchar();
+
+	return 0;
+}
+
+
+int main(void)
+{
+	int ret;
+	char *stack;
+
+	stack = (char *)malloc(1024 * sizeof(char));
+
+	graph_adapter_t adapter = GRAPH_ANY;
+	graph_mode_t mode = GRAPH_DEFMODE;
+	graph_freq_t freq = GRAPH_DEFFREQ;
+	graph_t graph;
+
+	rotrectangle_t rotrectangle;
+
+	rotrectangle.speed = 3;
+	rotrectangle.stop = 0;
+	rotrectangle.rec.a = 180;
+	rotrectangle.rec.b = 90;
+	rotrectangle.rec.color = 60000;
+
+	ret = rotrectangle_startpanel(&rotrectangle);
+
+	rotrectangle_switchmode(0);
```

Comment:
maybe switch mode only after all initialization has succeeded?

Aaand - it would be good to switch it back to the previous mode at exit (in case shell won't do it for You)